File: src/hangul.ts
export const HangulCharClass = {
  CompatibilityJamo: "CompatibilityJamo",
  JamoExtendedA: "JamoExtendedA",
  JamoExtendedB: "JamoExtendedB",
  Jamo: "Jamo",
  Syllables: "Syllables",
  None: "None",
} as const;

export type HangulCharClass =
  (typeof HangulCharClass)[keyof typeof HangulCharClass];

export const HangulCharClassStatic = {
  from(value: string): HangulCharClass {
    if (value.length === 0) return HangulCharClass.None;

    const codePoint = value.charCodeAt(0);

    if (codePoint >= 0xac00 && codePoint <= 0xd7af) {
      return HangulCharClass.Syllables;
    } else if (codePoint >= 0x1100 && codePoint <= 0x11ff) {
      return HangulCharClass.Jamo;
    } else if (codePoint >= 0x3130 && codePoint <= 0x318f) {
      return HangulCharClass.CompatibilityJamo;
    } else if (codePoint >= 0xa960 && codePoint <= 0xa97f) {
      return HangulCharClass.JamoExtendedA;
    } else if (codePoint >= 0xd7b0 && codePoint <= 0xd7ff) {
      return HangulCharClass.JamoExtendedB;
    }

    return HangulCharClass.None;
  },

  split(value: string): Array<[HangulCharClass, string]> {
    const result: Array<[HangulCharClass, string]> = [];
    let pos: { startIdx: number; class: HangulCharClass } | null = null;

    let currIdx = 0;
    for (const char of value) {
      const charClass = HangulCharClassStatic.from(char);

      if (pos) {
        if (charClass !== pos.class) {
          result.push([pos.class, value.substring(pos.startIdx, currIdx)]);
          pos = { startIdx: currIdx, class: charClass };
        }
      } else {
        pos = { startIdx: currIdx, class: charClass };
      }

      currIdx += char.length;
    }

    if (pos) {
      result.push([pos.class, value.substring(pos.startIdx)]);
    }

    return result;
  },
};

export function decompose_hangul_syllable_to_jamos(
  ch: string,
): [string, string, string?] | null {
  if (ch.length === 0) return null;

  const charClass = HangulCharClassStatic.from(ch);
  const codepoint = ch.charCodeAt(0);

  if (charClass !== HangulCharClass.Syllables) {
    return null;
  }

  const baseCodepoint = codepoint - 0xac00;
  const initialCodepointIdx = Math.floor(baseCodepoint / 588);
  const medialCodepointIdx = Math.floor(
    (baseCodepoint - initialCodepointIdx * 588) / 28,
  );
  const finalCodepointIdx =
    baseCodepoint - initialCodepointIdx * 588 - medialCodepointIdx * 28;

  const initialCodepoint = 0x1100 + initialCodepointIdx;
  const medialCodepoint = 0x1161 + medialCodepointIdx;
  const finalCodepoint = 0x11a7 + finalCodepointIdx;

  const initialCh = String.fromCharCode(initialCodepoint);
  const medialCh = String.fromCharCode(medialCodepoint);
  const maybeFinalCh =
    finalCodepointIdx === 0 ? undefined : String.fromCharCode(finalCodepoint);

  return [initialCh, medialCh, maybeFinalCh];
}

export function hangul_jamo_to_compat(ch: string): string | null {
  const jamoToCompatMap: { [key: string]: string } = {
    // Consonants
    ᄀ: "ㄱ",
    ᆨ: "ㄱ",
    ᄁ: "ㄲ",
    ᆩ: "ㄲ",
    ᆪ: "ㄳ",
    ᄂ: "ㄴ",
    ᆫ: "ㄴ",
    ᆬ: "ㄵ",
    ᆭ: "ㄶ",
    ᄃ: "ㄷ",
    ᆮ: "ㄷ",
    ᄄ: "ㄸ",
    ᄅ: "ㄹ",
    ᆯ: "ㄹ",
    ᆰ: "ㄺ",
    ᆱ: "ㄻ",
    ᆲ: "ㄼ",
    ᆳ: "ㄽ",
    ᆴ: "ㄾ",
    ᆵ: "ㄿ",
    ᆶ: "ㅀ",
    ᄆ: "ㅁ",
    ᆷ: "ㅁ",
    ᄇ: "ㅂ",
    ᆸ: "ㅂ",
    ᄈ: "ㅃ",
    ᆹ: "ㅄ",
    ᄉ: "ㅅ",
    ᆺ: "ㅅ",
    ᄊ: "ㅆ",
    ᆻ: "ㅆ",
    ᄋ: "ㅇ",
    ᆼ: "ㅇ",
    ᄌ: "ㅈ",
    ᆽ: "ㅈ",
    ᄍ: "ㅉ",
    ᄎ: "ㅊ",
    ᆾ: "ㅊ",
    ᄏ: "ㅋ",
    ᆿ: "ㅋ",
    ᄐ: "ㅌ",
    ᇀ: "ㅌ",
    ᄑ: "ㅍ",
    ᇁ: "ㅍ",
    ᄒ: "ㅎ",
    ᇂ: "ㅎ",

    // Vowels
    ᅡ: "ㅏ",
    ᅢ: "ㅐ",
    ᅣ: "ㅑ",
    ᅤ: "ㅒ",
    ᅥ: "ㅓ",
    ᅦ: "ㅔ",
    ᅧ: "ㅕ",
    ᅨ: "ㅖ",
    ᅩ: "ㅗ",
    ᅪ: "ㅘ",
    ᅫ: "ㅙ",
    ᅬ: "ㅚ",
    ᅭ: "ㅛ",
    ᅮ: "ㅜ",
    ᅯ: "ㅝ",
    ᅰ: "ㅞ",
    ᅱ: "ㅟ",
    ᅲ: "ㅠ",
    ᅳ: "ㅡ",
    ᅴ: "ㅢ",
    ᅵ: "ㅣ",
  };

  return jamoToCompatMap[ch] || null;
}

export function hangul_jamo_to_compat_with_fallback(ch: string): string {
  return hangul_jamo_to_compat(ch) || ch;
}

function hangul_syllable_to_jamos(ch: string): string | null {
  const decomposed = decompose_hangul_syllable_to_jamos(ch);
  if (!decomposed) return null;

  const [initial, medial, maybeFinal] = decomposed;
  return maybeFinal
    ? `${initial}${medial}${maybeFinal}`
    : `${initial}${medial}`;
}

export function decompose_all_hangul_syllables(value: string): string {
  let result = "";

  for (const ch of value) {
    const jamos = hangul_syllable_to_jamos(ch);
    if (jamos) {
      result += jamos;
    } else {
      result += ch;
    }
  }

  return result;
}

File: src/hangul.rs
#[derive(Debug, PartialEq, Copy, Clone)]
pub enum HangulCharClass {
    CompatibilityJamo,
    JamoExtendedA,
    JamoExtendedB,
    Jamo,
    Syllables,
    None,
}

impl From<char> for HangulCharClass {
    fn from(value: char) -> Self {
        match value {
            '\u{ac00}'..='\u{d7af}' => HangulCharClass::Syllables,
            '\u{1100}'..='\u{11ff}' => HangulCharClass::Jamo,
            '\u{3130}'..='\u{318f}' => HangulCharClass::CompatibilityJamo,
            '\u{a960}'..='\u{a97f}' => HangulCharClass::JamoExtendedA,
            '\u{d7b0}'..='\u{d7ff}' => HangulCharClass::JamoExtendedB,
            _ => HangulCharClass::None,
        }
    }
}

impl HangulCharClass {
    pub fn split(value: &str) -> Vec<(HangulCharClass, &str)> {
        let mut result = vec![];
        let mut pos: Option<(usize, HangulCharClass)> = None;
        for (curr_idx, char) in value.char_indices() {
            if let Some((start_idx, class)) = pos {
                if HangulCharClass::from(char) != class {
                    result.push((class, &value[start_idx..curr_idx]));
                    pos = Some((curr_idx, HangulCharClass::from(char)));
                }
            } else {
                pos = Some((curr_idx, HangulCharClass::from(char)));
            }
        }
        if let Some((start_idx, class)) = pos {
            result.push((class, &value[start_idx..]));
        }
        result
    }
}

/// Decomposes the given Hangul syllable into its
/// composite Hangul jamos.
///
/// If the character is not a Hangul syllable, returns
/// None.
pub fn decompose_hangul_syllable_to_jamos(ch: char) -> Option<(char, char, Option<char>)> {
    let class = HangulCharClass::from(ch);
    let codepoint = ch as u32;
    if class != HangulCharClass::Syllables {
        return None;
    }
    let base_codepoint = codepoint - 0xac00;
    let initial_codepoint_idx = base_codepoint / 588;
    let medial_codepoint_idx = (base_codepoint - (initial_codepoint_idx * 588)) / 28;
    let final_codepoint_idx =
        base_codepoint - (initial_codepoint_idx * 588) - (medial_codepoint_idx * 28);
    let initial_codepoint = 0x1100 + initial_codepoint_idx;
    let medial_codepoint = 0x1161 + medial_codepoint_idx;
    let final_codepoint = 0x11a7 + final_codepoint_idx;
    let initial_ch = char::from_u32(initial_codepoint).unwrap();
    let medial_ch = char::from_u32(medial_codepoint).unwrap();
    let maybe_final_ch = if final_codepoint_idx == 0 {
        None
    } else {
        char::from_u32(final_codepoint)
    };
    assert_eq!(HangulCharClass::from(initial_ch), HangulCharClass::Jamo);
    assert_eq!(HangulCharClass::from(medial_ch), HangulCharClass::Jamo);
    Some((initial_ch, medial_ch, maybe_final_ch))
}

/// Converts a Hangul Jamo to its equivalent
/// Hangul Compatibility Jamo.
///
/// This can be used when you want to display the
/// Jamo by itself, and ensure that it's displayed
/// without weird spacing on either side (which it seems
/// like terminals often do in inconsistent ways).
pub fn hangul_jamo_to_compat(ch: char) -> Option<char> {
    match ch {
        // Consonants
        'ᄀ' | 'ᆨ' => Some('ㄱ'),
        'ᄁ' | 'ᆩ' => Some('ㄲ'),
        'ᆪ' => Some('ㄳ'),
        'ᄂ' | 'ᆫ' => Some('ㄴ'),
        'ᆬ' => Some('ㄵ'),
        'ᆭ' => Some('ㄶ'),
        'ᄃ' | 'ᆮ' => Some('ㄷ'),
        'ᄄ' => Some('ㄸ'),
        'ᄅ' | 'ᆯ' => Some('ㄹ'),
        'ᆰ' => Some('ㄺ'),
        'ᆱ' => Some('ㄻ'),
        'ᆲ' => Some('ㄼ'),
        'ᆳ' => Some('ㄽ'),
        'ᆴ' => Some('ㄾ'),
        'ᆵ' => Some('ㄿ'),
        'ᆶ' => Some('ㅀ'),
        'ᄆ' | 'ᆷ' => Some('ㅁ'),
        'ᄇ' | 'ᆸ' => Some('ㅂ'),
        'ᄈ' => Some('ㅃ'),
        'ᆹ' => Some('ㅄ'),
        'ᄉ' | 'ᆺ' => Some('ㅅ'),
        'ᄊ' | 'ᆻ' => Some('ㅆ'),
        'ᄋ' | 'ᆼ' => Some('ㅇ'),
        'ᄌ' | 'ᆽ' => Some('ㅈ'),
        'ᄍ' => Some('ㅉ'),
        'ᄎ' | 'ᆾ' => Some('ㅊ'),
        'ᄏ' | 'ᆿ' => Some('ㅋ'),
        'ᄐ' | 'ᇀ' => Some('ㅌ'),
        'ᄑ' | 'ᇁ' => Some('ㅍ'),
        'ᄒ' | 'ᇂ' => Some('ㅎ'),

        // Vowels
        'ᅡ' => Some('ㅏ'),
        'ᅢ' => Some('ㅐ'),
        'ᅣ' => Some('ㅑ'),
        'ᅤ' => Some('ㅒ'),
        'ᅥ' => Some('ㅓ'),
        'ᅦ' => Some('ㅔ'),
        'ᅧ' => Some('ㅕ'),
        'ᅨ' => Some('ㅖ'),
        'ᅩ' => Some('ㅗ'),
        'ᅪ' => Some('ㅘ'),
        'ᅫ' => Some('ㅙ'),
        'ᅬ' => Some('ㅚ'),
        'ᅭ' => Some('ㅛ'),
        'ᅮ' => Some('ㅜ'),
        'ᅯ' => Some('ㅝ'),
        'ᅰ' => Some('ㅞ'),
        'ᅱ' => Some('ㅟ'),
        'ᅲ' => Some('ㅠ'),
        'ᅳ' => Some('ㅡ'),
        'ᅴ' => Some('ㅢ'),
        'ᅵ' => Some('ㅣ'),

        _ => None,
    }
}

/// Converts a Hangul Jamo to its equivalent
/// Hangul Compatibility Jamo.
///
/// If there isn't a corresponding one, it just returns
/// the original character unchanged.
pub fn hangul_jamo_to_compat_with_fallback(ch: char) -> char {
    hangul_jamo_to_compat(ch).unwrap_or(ch)
}

fn hangul_syllable_to_jamos(ch: char) -> Option<String> {
    if let Some((initial_ch, medial_ch, maybe_final_ch)) = decompose_hangul_syllable_to_jamos(ch) {
        if let Some(final_ch) = maybe_final_ch {
            Some(format!("{initial_ch}{medial_ch}{final_ch}"))
        } else {
            Some(format!("{initial_ch}{medial_ch}"))
        }
    } else {
        None
    }
}

/// Converts any Hangul syllables in the given string into
/// Hangul jamos.
pub fn decompose_all_hangul_syllables<T: AsRef<str>>(value: T) -> String {
    let str = value.as_ref();
    let mut result = String::with_capacity(str.len());

    for ch in str.chars() {
        if let Some(jamos) = hangul_syllable_to_jamos(ch) {
            result.push_str(&jamos);
        } else {
            result.push(ch);
        }
    }

    result
}

#[cfg(test)]
mod test {
    use crate::hangul::{
        decompose_all_hangul_syllables, decompose_hangul_syllable_to_jamos, HangulCharClass,
    };

    #[test]
    fn test_char_class_works() {
        assert_eq!(HangulCharClass::from('이'), HangulCharClass::Syllables);
        assert_eq!(HangulCharClass::from('ᆸ'), HangulCharClass::Jamo);
        assert_eq!(
            HangulCharClass::from('ㄱ'),
            HangulCharClass::CompatibilityJamo
        );
    }

    #[test]
    fn test_decompose_works() {
        assert_eq!(decompose_hangul_syllable_to_jamos('h'), None);
        assert_eq!(
            decompose_hangul_syllable_to_jamos('이'),
            Some(('ᄋ', 'ᅵ', None))
        );
        assert_eq!(
            decompose_hangul_syllable_to_jamos('는'),
            Some(('ᄂ', 'ᅳ', Some('ᆫ')))
        );
    }

    #[test]
    fn test_decompose_all_works() {
        let orig = "이";
        assert_eq!(orig.chars().count(), 1);
        let decomposed = "이";
        assert_eq!(decomposed.chars().count(), 2);
        assert_eq!(decompose_all_hangul_syllables(&orig), decomposed.to_owned());
    }

    #[test]
    fn test_split_works() {
        assert_eq!(HangulCharClass::split(""), vec![]);
        assert_eq!(
            HangulCharClass::split("이"),
            vec![(HangulCharClass::Syllables, "이")]
        );

        assert_eq!(
            HangulCharClass::split("hi 이 there"),
            vec![
                (HangulCharClass::None, "hi "),
                (HangulCharClass::Syllables, "이"),
                (HangulCharClass::None, " there")
            ]
        );
    }
}
